Incorrect upper bound on type parameter with inner classes

Consider this test case:
```
import org.checkerframework.checker.nullness.qual.Nullable;

public class TestTypeParamsCons<T extends @Nullable Object> {
    public class C {
        public C(T a) {}

        public void f(T a) {
            new C(a);
        }
    }
}
```

The constructor for `C` expects a `T`, which is exactly what is passed in `f`. These should both refer to the same type, `T extends @Nullable Object`. When I run the nullness checker with the command ` javac -processor nullness TestTypeParamsCons.java`, I get the following error:

```
./checker/tests/nullness/TestTypeParamsCons.java:8: error: [argument.type.incompatible] incompatible types in argument.
            new C(a);
                  ^
  found   : T extends @Initialized @Nullable Object
  required: T extends @Initialized @NonNull Object
1 error

```

--- a/checker/tests/nullness/Issue2865.java
+++ b/checker/tests/nullness/Issue2865.java
@@ -17,4 +17,9 @@ public class Issue2865<T extends @Nullable Object> {
         s.new C(null);
         s.new C("");
     }
+
+    void test2(Issue2865<@Nullable String> s) {
+        s.new C(null);
+        s.new C("");
+    }
 }

--- a/checker/tests/nullness/Issue2865.java
+++ b/checker/tests/nullness/Issue2865.java
@@ -1,0 +1,20 @@
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class Issue2865<T extends @Nullable Object> {
+    public class C {
+        public C(T a) {}
+
+        public void f(T a) {
+            new C(a);
+            // :: error: (argument.type.incompatible)
+            new C(null);
+        }
+    }
+
+    void test(Issue2865<@NonNull String> s) {
+        // :: error: (argument.type.incompatible)
+        s.new C(null);
+        s.new C("");
+    }
+}

--- a/framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java
@@ -2113,12 +2113,19 @@ public class AnnotatedTypeFactory implements AnnotationProvider {
      * newClassTree has a diamond operator. In that case, the annotations on the type arguments are
      * inferred using the assignment context and contain defaults.
      *
+     * <p>Also, fully annotates the enclosing type of the returned declared type.
+     *
      * <p>(Subclass beside {@link GenericAnnotatedTypeFactory} should not override this method.)
      *
      * @param newClassTree NewClassTree
      * @return AnnotatedDeclaredType
      */
     public AnnotatedDeclaredType fromNewClass(NewClassTree newClassTree) {
+
+        AnnotatedDeclaredType enclosingType = null;
+        if (newClassTree.getEnclosingExpression() != null) {
+            enclosingType = (AnnotatedDeclaredType) getReceiverType(newClassTree);
+        }
         // Diamond trees that are not anonymous classes.
         if (TreeUtils.isDiamondTree(newClassTree) && newClassTree.getClassBody() == null) {
             AnnotatedDeclaredType type =
@@ -2146,6 +2153,7 @@ public class AnnotatedTypeFactory implements AnnotationProvider {
                     (AnnotatedDeclaredType)
                             TypeFromTree.fromTypeTree(this, newClassTree.getIdentifier());
             type.replaceAnnotations(fromTypeTree.getAnnotations());
+            type.setEnclosingType(enclosingType);
             return type;
         } else if (newClassTree.getClassBody() != null) {
             AnnotatedDeclaredType type =
@@ -2156,13 +2164,17 @@ public class AnnotatedTypeFactory implements AnnotationProvider {
             List<? extends AnnotationTree> annos =
                     newClassTree.getClassBody().getModifiers().getAnnotations();
             type.addAnnotations(TreeUtils.annotationsFromTypeAnnotationTrees(annos));
+            type.setEnclosingType(enclosingType);
             return type;
         } else {
             // If newClassTree does not create anonymous class,
             // newClassTree.getIdentifier includes the explicit annotations in this location:
             //   new @HERE Class()
-            return (AnnotatedDeclaredType)
-                    TypeFromTree.fromTypeTree(this, newClassTree.getIdentifier());
+            AnnotatedDeclaredType type =
+                    (AnnotatedDeclaredType)
+                            TypeFromTree.fromTypeTree(this, newClassTree.getIdentifier());
+            type.setEnclosingType(enclosingType);
+            return type;
         }
     }
 

--- a/framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java
@@ -2122,9 +2122,11 @@ public class AnnotatedTypeFactory implements AnnotationProvider {
      */
     public AnnotatedDeclaredType fromNewClass(NewClassTree newClassTree) {
 
-        AnnotatedDeclaredType enclosingType = null;
+        AnnotatedDeclaredType enclosingType;
         if (newClassTree.getEnclosingExpression() != null) {
             enclosingType = (AnnotatedDeclaredType) getReceiverType(newClassTree);
+        } else {
+            enclosingType = null;
         }
         // Diamond trees that are not anonymous classes.
         if (TreeUtils.isDiamondTree(newClassTree) && newClassTree.getClassBody() == null) {

--- a/framework/src/main/java/org/checkerframework/framework/util/typeinference/TypeArgInferenceUtil.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/typeinference/TypeArgInferenceUtil.java
@@ -181,6 +181,10 @@ public class TypeArgInferenceUtil {
         } else if (assignmentContext instanceof NewClassTree) {
             // This need to be basically like MethodTree
             NewClassTree newClassTree = (NewClassTree) assignmentContext;
+            if (newClassTree.getEnclosingExpression() instanceof NewClassTree
+                    && (newClassTree.getEnclosingExpression() == path.getLeaf())) {
+                return null;
+            }
             ExecutableElement constructorElt = TreeUtils.constructor(newClassTree);
             AnnotatedTypeMirror receiver = atypeFactory.fromNewClass(newClassTree);
             res =

--- a/javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java
@@ -745,44 +745,44 @@ public final class TreeUtils {
 
     /** Returns the receiver tree of a field access or a method invocation. */
     public static ExpressionTree getReceiverTree(ExpressionTree expression) {
-        ExpressionTree receiver = TreeUtils.withoutParens(expression);
+        ExpressionTree receiver;
+        switch (expression.getKind()) {
+            case METHOD_INVOCATION:
+                // Trying to handle receiver calls to trees of the form
+                //     ((m).getArray())
+                // returns the type of 'm' in this case
+                receiver = ((MethodInvocationTree) expression).getMethodSelect();
 
-        if (!(receiver.getKind() == Tree.Kind.METHOD_INVOCATION
-                || receiver.getKind() == Tree.Kind.MEMBER_SELECT
-                || receiver.getKind() == Tree.Kind.IDENTIFIER
-                || receiver.getKind() == Tree.Kind.ARRAY_ACCESS)) {
-            // No receiver tree for anything but these four kinds.
-            return null;
+                if (receiver.getKind() == Tree.Kind.IDENTIFIER) {
+                    // It's a method call "m(foo)" without an explicit receiver
+                    return null;
+                } else if (receiver.getKind() == Tree.Kind.MEMBER_SELECT) {
+                    receiver = ((MemberSelectTree) receiver).getExpression();
+                } else {
+                    // Otherwise, e.g. a NEW_CLASS: nothing to do.
+                    return null;
+                }
+                break;
+            case NEW_CLASS:
+                receiver = ((NewClassTree) expression).getEnclosingExpression();
+                break;
+            case ARRAY_ACCESS:
+                receiver = ((ArrayAccessTree) expression).getExpression();
+                break;
+            case MEMBER_SELECT:
+                receiver = ((MemberSelectTree) expression).getExpression();
+                // Avoid int.class
+                if (receiver instanceof PrimitiveTypeTree) {
+                    return null;
+                }
+                break;
+            case IDENTIFIER:
+                // It's a field access on implicit this or a local variable/parameter.
+                return null;
+            default:
+                return null;
         }
 
-        if (receiver.getKind() == Tree.Kind.METHOD_INVOCATION) {
-            // Trying to handle receiver calls to trees of the form
-            //     ((m).getArray())
-            // returns the type of 'm' in this case
-            receiver = ((MethodInvocationTree) receiver).getMethodSelect();
-
-            if (receiver.getKind() == Tree.Kind.IDENTIFIER) {
-                // It's a method call "m(foo)" without an explicit receiver
-                return null;
-            } else if (receiver.getKind() == Tree.Kind.MEMBER_SELECT) {
-                receiver = ((MemberSelectTree) receiver).getExpression();
-            } else {
-                // Otherwise, e.g. a NEW_CLASS: nothing to do.
-            }
-        } else if (receiver.getKind() == Tree.Kind.IDENTIFIER) {
-            // It's a field access on implicit this or a local variable/parameter.
-            return null;
-        } else if (receiver.getKind() == Tree.Kind.ARRAY_ACCESS) {
-            return TreeUtils.withoutParens(((ArrayAccessTree) receiver).getExpression());
-        } else if (receiver.getKind() == Tree.Kind.MEMBER_SELECT) {
-            receiver = ((MemberSelectTree) receiver).getExpression();
-            // Avoid int.class
-            if (receiver instanceof PrimitiveTypeTree) {
-                return null;
-            }
-        }
-
-        // Receiver is now really just the receiver tree.
         return TreeUtils.withoutParens(receiver);
     }
 

--- a/javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java
@@ -753,13 +753,10 @@ public final class TreeUtils {
                 // returns the type of 'm' in this case
                 receiver = ((MethodInvocationTree) expression).getMethodSelect();
 
-                if (receiver.getKind() == Tree.Kind.IDENTIFIER) {
-                    // It's a method call "m(foo)" without an explicit receiver
-                    return null;
-                } else if (receiver.getKind() == Tree.Kind.MEMBER_SELECT) {
+                if (receiver.getKind() == Tree.Kind.MEMBER_SELECT) {
                     receiver = ((MemberSelectTree) receiver).getExpression();
                 } else {
-                    // Otherwise, e.g. a NEW_CLASS: nothing to do.
+                    // It's a method call "m(foo)" without an explicit receiver
                     return null;
                 }
                 break;

